Invalidate sandbox FieldValues cache on question change

metamben opened the pull request:
Fixes #25411.

This PR includes the last modification date of the question (`report_card.update_at`) in the list of fields used for checking for validating the cache.

As it stands, this PR creates multiple `FieldValues` entries of type `:sandbox` with different `hash_key`s for the same field which, according to #668, might be wrong.

metamben committed:
Invalidate sandbox FieldValues cache on question change

See #25411 for context.

metamben committed:
Wait 1 second if the application DB is mysql

metamben committed:
Always sleep a second

metamben committed:
Merge branch 'master' into 25411-invalidate-sandbox-field-values-on-question-change

metamben commented on enterprise/backend/test/metabase_enterprise/sandbox/models/params/field_values_test.clj:
```diff
@@ -45,7 +47,23 @@
+            (let [new-query (mt/mbql-query categories
+                                           {:filter [:and [:> $id 1] [:< $id 4]]})]
+              ;; sleeping should ensure that updated_at changes
+              (Thread/sleep 1000)
```
MySQL seems to have second granularity, which makes this test somewhat slow. This also means that if the sandboxed user uses the sandboxed query within a second after the modification of the query, they will. get wrong filter values. I don't think this is a problem in practice.

camsaul replied:
Yeah, `timestamp` in MySQL is second granularity by default. You should just add a migration to change that column to `timestamp(3)` (millisecond) or `timestamp(6)` (nanosecond) so we don't have to worry about this in the future. We've done it a bunch of times in the past to fix the same situation with other columns, so there's prior art in the migration file that you could refer to

metamben replied:
I suppose `timestamp_type` (microseconds) is the right type to set. I'll update the PR.

metamben committed:
Merge remote-tracking branch 'origin/master' into 25411-invalidate-sandbox-field-values-on-question-change

metamben committed:
Increase precision of report_card.updated_at to a microsecond

This is so that the test doesn't have to wait a second to get a modified timestamp.
It's also more consistent with other timestamp columns.

metamben committed:
Add back test cases for :linked-filter

metamben committed:
Merge branch 'master' into 25411-invalidate-sandbox-field-values-on-question-change

metamben commented on resources/migrations/000_migrations.yaml:
```diff
@@ -13990,6 +13990,21 @@ databaseChangeLog:
             constraintName: fk_action_creator_id
             nullable: false
 
+  - changeSet:
+      id: v46.00-074
+      author: metamben
+      comment: Added 0.46.0 -- increase precision of updated_at of report_card
+      changes:
+        - modifyDataType:
+            tableName: report_card
+            columnName: updated_at
+            newDataType: ${timestamp_type}
+      rollback:
+        - modifyDataType:
+            tableName: report_card
+            columnName: updated_at
+            newDataType: DATETIME
+
```
Should this change be constrained to mysql and mariadb?

camsaul replied:
I think it makes sense for Postgres and H2 too, since `${timestamp_type}` is `TIMESTAMP FOR TIME ZONE` for those, which is definitely preferrable to `DATETIME` for storing timestamps

metamben committed:
Merge remote-tracking branch 'origin/master' into 25411-invalidate-sandbox-field-values-on-question-change

metamben committed:
Fix migration ordering
